### PR TITLE
Boosted forest

### PR DESCRIPTION
**Pull request by hectorgon**
Boosted decision trees made scalable by importance sampling from the data set while building trees
then mini batch gradient descent in between.

@deerzq @tension7 @sdemars

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            input : RDD[Example],
+            config : Config,
+            key : String) : ForestModel = {
+    val candidateSize : Int = config.getInt(key + ".num_candidates")
+    val rankKey : String = config.getString(key + ".rank_key")
+    val rankThreshold : Double = config.getDouble(key + ".rank_threshold")
+    val maxDepth : Int = config.getInt(key + ".max_depth")
+    val minLeafCount : Int = config.getInt(key + ".min_leaf_items")
+    val numTries : Int = config.getInt(key + ".num_tries")
+    val splitCriteria : String = Try(config.getString(key + ".split_criteria")).getOrElse("gini")
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: A general comment not for this specific PR: we should refactor the config loading part to something like taskConfig = config.getConfig(key) then, param = taskConfig.getString(param_name).

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                  input : RDD[Example],
+                  config : Config,
+                  key : String,
+                  opt : BoostedForestTrainerOptions) = {
+     for (i <- 0 until opt.iterations) {
+       log.info("Running boost iteration %d".format(i))
+       val forestBC = sc.broadcast(forest)
+       // Get forest responses
+       val labelSumResponse = LinearRankerUtils
+              .makePointwiseFloat(input, config, key)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: perhaps we also refactor 'makePointwiseFloat' this to TrainingUtils

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+import scala.collection.JavaConverters._
+
+// A boosted tree forest trainer.
+// Alternates between fitting a tree and building one on the importance
+// sampled outliers of previous trees.
+object BoostedForestTrainer {
+  private final val log: Logger = LoggerFactory.getLogger("BoostedForestTrainer")
+  
+  case class BoostedForestTrainerOptions(candidateSize : Int,
+                                         rankKey : String,
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: a nit about naming: `Options` sounds confusing with with Option in scala

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+       var sum : Double = 0.0
+       for (cg <- countAndGradient) {
+         val key = cg._1
+         val tree = trees(key._1)
+         val stump = tree.getStumps().get(key._2)
+         val (count, grad) = cg._2
+         val curr = stump.getFeatureWeight()
+         val avgGrad = grad / count
+         stump.setFeatureWeight(curr - opt.learningRate * avgGrad)
+         sum = avgGrad * avgGrad
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: `sum += avgGrad * avgGrad` ?

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.util.Random
+import scala.util.Try
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+// A boosted tree forest trainer.
+// Alternates between fitting a tree and building one on the importance
+// sampled outliers of previous trees.
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: add some reference to explain your implementation

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala (hunk 1 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: Done

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala (hunk 3 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: Done

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala (hunk 4 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: Done

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala (hunk 5 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: Done

File: training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala (hunk 2 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/BoostedForestTrainer.scala]: Ok will do in a later PR

**Comment by deerzq** [comment]
LGTM

**Pull request merged**